homeassistant/components/sma/__init__.py

```diff
@@ -1 +1,199 @@
+    to support deprecated yaml config from platform setup.
+    """
+
+    # Add sensors from the custom config
+    sensor_def.add(
+        [
+            pysma.Sensor(o[CONF_KEY], n, o[CONF_UNIT], o[CONF_FACTOR], o.get(CONF_PATH))
+            for n, o in entry.data.get(CONF_CUSTOM).items()
+        ]
+    )
+
+    # Parsing of sensors configuration
+    config_sensors = entry.data.get(CONF_SENSORS)
+    if not config_sensors:
+        return
+
+    # Find and replace sensors removed from pysma
+    # This only alters the config, the actual sensor migration takes place in _migrate_old_unique_ids
+    for sensor in config_sensors.copy():
+        if sensor in pysma.LEGACY_MAP:
+            config_sensors.remove(sensor)
+            config_sensors.append(pysma.LEGACY_MAP[sensor]["new_sensor"])
+
+    # Only sensors from config should be enabled
+    for sensor in sensor_def:
+        sensor.enabled = sensor.name in config_sensors
+
+
+async def _migrate_old_unique_ids(
+    hass: HomeAssistant, entry: ConfigEntry, sensor_def: pysma.Sensors
+) -> None:
+    """Migrate legacy sensor entity_id format to new format."""
+    entity_registry = er.async_get(hass)
+
+    # Create list of all possible sensor names
+    possible_sensors = list(
+        set(
+            entry.data.get(CONF_SENSORS)
+            + [s.name for s in sensor_def]
+            + list(pysma.LEGACY_MAP)
+        )
+    )
+
+    for sensor in possible_sensors:
+        if sensor in sensor_def:
+            pysma_sensor = sensor_def[sensor]
+            original_key = pysma_sensor.key
+        elif sensor in pysma.LEGACY_MAP:
+            # If sensor was removed from pysma we will remap it to the new sensor
+            legacy_sensor = pysma.LEGACY_MAP[sensor]
+            pysma_sensor = sensor_def[legacy_sensor["new_sensor"]]
+            original_key = legacy_sensor["old_key"]
+        else:
+            _LOGGER.error("%s does not exist", sensor)
+            continue
+
+        # Find entity_id using previous format of unique ID
+        entity_id = entity_registry.async_get_entity_id(
+            "sensor", "sma", f"sma-{original_key}-{sensor}"
+        )
+
+        if not entity_id:
+            continue
+
+        # Change entity_id to new format using the device serial in entry.unique_id
+        new_unique_id = f"{entry.unique_id}-{pysma_sensor.key}_{pysma_sensor.key_idx}"
+        entity_registry.async_update_entity(entity_id, new_unique_id=new_unique_id)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up sma from a config entry."""
+    # Init all default sensors
+    sensor_def = pysma.Sensors()
+
+    if entry.source == SOURCE_IMPORT:
+        await _parse_legacy_options(entry, sensor_def)
+        await _migrate_old_unique_ids(hass, entry, sensor_def)
+
+    # Init the SMA interface
+    protocol = "https" if entry.data.get(CONF_SSL) else "http"
+    url = f"{protocol}://{entry.data.get(CONF_HOST)}"
```

Comment:
Please don't use dict.get when we know that the key is in the dict. Use `dict[key]`.

Reply:
I'll change these to `dict[key]`. Any specific reason why?

Reply:
The primary reason: If we think the key exists in the dict and are mistaken we'll get a KeyError and know about our mistake. `dict.get` would swallow our mistake. It's also more clear for readers of the code what the intention of the code writer is, if following this practice.

Secondary: `dict[key]` is faster than `dict.get(key)`.